### PR TITLE
Issue/lims 1536 Add buttons for reference widgets dynamically

### PR DESCRIPTION
**Pull request by xispa**
**This other PR should be reviewed and accepted or rejected at the same time**: http://git.io/tb3cjQ

**Basic usage**
```python
ExtReferenceField(
            'Doctor',
            # ... other stuff ....
            widget=ReferenceWidget(
               # ... other stuff ...
                add_button={
                    'visible': True,
                    'url': 'doctors/portal_factory/Doctor/new/edit',
                    'return_fields': ['Firstname', 'Surname'],
                }
            ),
        ),
```

url: the url to be loaded inside the overlay
return_fields: the fields to be used to fill the referencewidget after the submit

**Advanced usage**
```python
ExtReferenceField(
            'Patient',
            # ... other stuff ....
            widget=ReferenceWidget(
                # ... other stuff ....
                add_button={
                    'visible': True,
                    'url': 'patients/portal_factory/Patient/new/edit',
                    'return_fields': ['Firstname', 'Surname'],
                    'js_controllers': ['#patient-base-edit',],
                    'overlay_handler': 'HealthPatientOverlayHandler',
                }
            ),
        ),
```
js_controllers: the controllers we need to load specifically for the view to be loaded inside the overlay
In this case, a specific js handler is required (overlay_handler) to manage the overlay event, so we need to add it:
https://github.com/xispa/bika.health/blob/issue/HEALTH-136/bika/health/static/js/bika.health.patient.js#L440-L472
Easy!

**Expert usage**
```javascript
add_button={
    'visible': True,
    'url': 'patients/portal_factory/Patient/new/edit',
    'return_fields': ['Firstname', 'Surname'],
    'js_controllers': ['#patient-base-edit',],
    'overlay_handler': 'HealthPatientOverlayHandler',
    'overlay_options': {
        'filter': 'head>*,content>*:not(div.configlet),dl.portalMessage.error,dl.portalMessage.info',
        'formselector': 'form[id$="base-edit"]',
        'closeselector': '[name="form.button.cancel"]',
        'width': '70%',
        'noform': 'close',
    },
```
}

### CODE REVIEW

**Comment by xispa** [comment]
**NB** Although the examples here refers to the health extension, the same procedure applies for any Content Type, from Bika LIMS itself or any extension. Easy

File: bika/lims/browser/js/bika.lims.loader.js [hunk 1]
```diff
@@ -166,19 +168,30 @@ window.bika.lims.controllers =  {
...
+ * in the current view, but you want to be loaded anyway. To deal with overlay
+ * widgets, for example.
+ */
+window.bika.lims.loadControllers = function(all, controllerKeys) {
     var controllers = window.bika.lims.controllers;
+    var prev = _bika_lims_loaded_js.length;
     for (var key in controllers) {
-        if ($(key).length) {
+        // Check if the key have value. Also check if this key exists in the controllerKeys' array
+        // Doing that, you can load non called variables in the web page
```
Comment (rockfruit) [inline bika/lims/browser/js/bika.lims.loader.js]: Not sure about this comment - can you expand it for me?  I understand what you've done below, but I don't know what "load non called variables" means.

**Comment by rockfruit** [comment]
That is beautiful, that is.

File: bika/lims/browser/js/bika.lims.loader.js (hunk 1 above)
Comment (xispa) [inline bika/lims/browser/js/bika.lims.loader.js]: ummm... don't know. Has nosense to me... to be removed.

Comment (Espurna) [inline bika/lims/browser/js/bika.lims.loader.js]: I wanted to say that calling the function "loadControllers(false, [array with JS controllers keys from window.bika.lims.controllers])",allows you to force bika to load/reload JS controllers defined inside the array.

Maybe I have to change the comment to make it more clear.

Comment (xispa) [inline bika/lims/browser/js/bika.lims.loader.js]: Aahh, please change the comment at will. Thanks!

Comment (rockfruit) [inline bika/lims/browser/js/bika.lims.loader.js]: Now it makes perfect sense - please yes, add some extra words there for
posterity.  This is nice code, it deserves to be fully understood

**Pull request merged**